### DISCUSSION TITLE
Debugging a PDM-managed project in VSCode

### DISCUSSION THREAD

**Discussion by pawamoy** [pdm-project/pdm#376] [Show and tell]
I had a bit of trouble setting up VSCode to be able to debug my code and add breakpoints onto the libraries installed in `__pypackages__`, so here is a how to.

First, know that I installed PDM with pipx:

```bash
pipx install pdm
```

It used my system Python, which is `/usr/bin/python`, and is version 3.9.
If I wanted to use another version, I would have installed it like this:

```bash
pipx install --python python3.8 pdm
```

Then I set up my project to use Python 3.9 and installed the libraries:

```bash
cd myproject
pdm use -f 3.9
pdm install -d
```

My code lives in the `src` folder.

In VSCode, I defined the Python interpreter to `/usr/bin/python`.
I then added a Run configuration which sets the `PYTHONPATH` environment variable so Python can find both my code and the libraries in `__pypackages__`.

`.vscode/settings.json`

```json
{
    "python.pythonPath": "/usr/bin/python"
}
```

`.vscode/launch.json`

```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "pytest",
            "type": "python",
            "request": "launch",
            "module": "pytest",
            "args": ["-c", "config/pytest.ini", "-n", "auto", "tests"],
            "justMyCode": false,
            "env": {"PYTHONPATH": "src:__pypackages__/3.9/lib"}
        }
    ]
}
```

If you set up a different Python interpreter, make sure you set the `env` accordingly. For example with a Python 3.8 interpreter, add `__pypackages__/3.8/lib` to the Python path.

**Comment by sleiner** [comment]
Thanks for the showcase! :-) After setting up VS Code for PDM [as described in the current docs](https://pdm.fming.dev/latest/#vscode), I did not have to use any additional setup to get debugging and even running tests from the sidebar working. So no need for specialized launch configurations :-)

**Reply by pawamoy** [reply]
You're right, my hack above should not be needed anymore 🎉 
Thanks for the feedback!

**Reply by Hnasar** [reply]
I see your tip of using `env` is mentioned in the docs under the [_Cannot enable PEP582 globally?_](https://pdm.fming.dev/latest/#vscode) heading.

What I ended up doing was, to set this globally, rather than in each project's `launch.json`, I added this to `settings.json`:
```json
    "python.envFile": "${userHome}/.vscode/pyenv",
```
and in that `pyenv` file:
```ini
PYTHONPATH=c:\users\<my username here>\appdata\roaming\pdm\venv\lib\site-packages\pdm\pep582
```

(which is the `PYTHONPATH` set by `pdm --pep582` in _System Properties -> Advanced -> Environment Variables_ (using Windows))

But indeed, once I enabled _PEP-582_ globally then this option wasn't needed either! 🎉
